### DISCUSSION TITLE
How to run/use the preview generator?

### DISCUSSION THREAD

**Discussion by foxhoundv** [nextcloud/all-in-one#542] [Questions] [answered]
How does one edit the crontab for apps like the preview generator that need a job ran every # of minutes/hours/days?

@szaimen Thanks for referring me here.

**Comment by szaimen** [comment, marked as answer]
The docu was moved to https://github.com/nextcloud/all-in-one/discussions/2810

**Reply by foxhoundv** [reply]
Thank you, I would not have gotten step 2 (3rd one in the list) correctly. This helps greatly.

**Reply by KOCTET-SPB** [reply]
> add a new line to this file if not already present: */10 * * * * docker ps --format "{{.Names}}" | grep -q "^nextcloud-aio-nextcloud$" && docker exec nextcloud-aio-nextcloud php occ preview:pre-generate

This method didn't work.
Found on reddit.com:
*/10 * * * * docker exec --user www-data --tty nextcloud-aio-nextcloud php occ preview:pre-generate >> somelog.txt 2>&1 ,
required --tty, -it flag didn't work for me

**Reply by szaimen** [reply]
there is no -it anymore. also --user www-data was indeed missing and was re-added now.

**Reply by KOCTET-SPB** [reply]
Thanks, it works great now

**Reply by pepper-lees** [reply]
What if I'm using Windows? Should I use Windows' task scheduler instead?

**Reply by szaimen** [reply]
> What if I'm using Windows? Should I use Windows' task scheduler instead?

Yes. Apart from that is the preview generator not necessarily needed anymore with Nextcloud 26, imho.

**Reply by pepper-lees** [reply]
> > What if I'm using Windows? Should I use Windows' task scheduler instead?
> 
> Yes. Apart from that is the preview generator not necessarily needed anymore with Nextcloud 26, imho.

Oh, why is that? I'm pretty sure I'm running Nextcloud 26, but scrolling through my list of images was extremely slow (i.e. images took ages to load) until I installed the preview generator and ran the generate-all script.

**Reply by szaimen** [reply]
the concurrency of preview generation is limited starting with nc26 to not bring the instance down. So yeah technically the problem that the instance crashes when doing that is solved.

However indeed depending on the hardware and if you scroll fast it might not be the fastest to generate them on the fly.

**Reply by pepper-lees** [reply]
> the concurrency of preview generation is limited starting with nc26 to not bring the instance down. So yeah technically the problem that the instance crashes when doing that is solved.

Hmm. The documentation of previewgenerator isn't super clear to me on what the `preview:pre-generate` does under the hood. Does it look at a list of currently-unanalyzed files and then process them? If (i'm pulling numbers out of the air here) say, the max number of images it can process is 10, and there are 50 new images that need to be processed, will the `preview:pre-generate` job simply need to be run multiple times until it gets through the queue? (in this example, if it's set to run every 10 minutes, then would that mean it takes 50 minutes to get through the queue?) - or does it attempt to generate previews for _everything_ whenever it's run?

**Reply by szaimen** [reply]
Basically the idea is that you run `preview:generate-all` once which creates previews for all files and then `preview:pre-generate` in a cronjob which detects changed and added files and generates previews for these.

**Reply by hoggatt** [reply]
Thank you for the post! This is basically essential for the memories application.

**Reply by meichthys** [reply]
@szaimen Sorry to resurrect an old thread. Is intended that we use the Docker host's crontab? This doesn't seem ideal since it would break things if the docker containers were moved to a different host. If possible, i would prefer to use a crontab inside the nextcloud container, however i'm not sure if that's supported, or if it would be reasonable to make the crontab persistent when the image is updated.

**Reply by NiCyph** [reply]
@meichthys If you want to do this insider the container, you can add this to the Dockerfile and then build the image

`FROM nextcloud:apache`
.....
`RUN echo '*/6 * * * * php -f /var/www/html/occ preview:pre-generate' >> /var/spool/cron/crontabs/ww>`

**Comment by philiprenich** [comment]
Like @meichthys , I'm curious why this is done on the Docker host and not within a container. I don't know what the lifecycle issue would be of manually editing a crontab in a container that will be updated when new AIO versions are released though. Are there improvements to this process we can make?
